text/0036-consensus-memory-usage.md

```diff
@@ -0,0 +1,130 @@
+
+Signatures older than that should never be trusted: The signers could have unbonded and sold their tokens, so they'd have nothing to lose. They could have sold their secret keys or decided to attack the network, knowing they cannot be slashed anymore, and the keys could have been used to sign a fork of the blockchain ("long range attack"). That is why joining nodes must always start from a trusted hash that is no older than `B` eras.
+
+However, even within that period, signatures can only be trusted if we actually _do_ detect equivocations and slash validators. That's why the current implementation keeps the `B` most recent eras in memory and keeps syncing their protocol state: If someone attempted to mislead a joining node with a sufficiently recent trusted hash, they would have to equivocate in a recent era, and their equivocating units would be gossiped across the network and result in slashing.
+
+But it gets worse: If we are in era `E`, not only does the above require us to keep in memory all eras back to `E - B`. It also means that any unit can depend on evidence from up to `B` eras ago. So era `E - B` could contain a unit referring to evidence of an equivocation in era `E - 2 * B`, and we still need to be able to validate that evidence. Fortunately evidence does not have any dependencies of its own, so the eras before `E - B` can be almost empty. ([HWY-257](https://casperlabs.atlassian.net/browse/HWY-257) will clear them.)
+
+In a test with 75 validators, one block per minute, and an unbonding period of ~6 hours, this resulted in 310 MB of memory usage. HWY-257 should cut that in half, but it is still proportional to the unbonding period and to the _square_ of the number of validators. With 200 validators and an unbonding period of 1 day, that would already mean 4.4 GB of memory usage for consensus alone. And that is in the _good_ case, without any attackers trying to spam the protocol state.
+
+In this CEP we propose a way to hold all still-bonded validators accountable without the need to remember all those old protocol states, greatly reducing the memory requirement.
+
+
+## Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+We drop an era's protocol state from memory as soon as all its blocks have a sufficient amount of finality signatures, and consider those signatures a replacement for the protocol state: just like the state, they are a proof of finality.
+
+Conflicting signatures cause the signer to get slashed, so that — just like with equivocations in the Highway protocol state — the finality signatures are actually backed by the stake.
+
+If an attacker creates an equivocation in an old era's Highway protocol, either it will have no effect, because everyone has already received the signed blocks and dropped the protocol state; or someone who still has the protocol state will detect the equivocation and still cause the attacker to get slashed.
```

Comment:
Does it mean we need to detect equivocation between finality signature and the conflicting unit (consensus message)?

Reply:
I was wondering about that as well and [on Wednesday we kind of agreed that we don't](https://casperlabs.atlassian.net/wiki/spaces/EN/pages/250282071/Consensus+design+meetings#2021-02-24).

Basically: If the validators reach consensus anyway, without using Highway for it, then we can't really complain — they _have_ reached consensus and signed a consistent blockchain, after all. :man_shrugging: 
(Anyway, not sure about this — let me know your thoughts!)

Reply:
Although, now that I think of it: It means that we lose the security guarantee for clients following the Highway protocol at a custom fault tolerance threshold. Seeing a block finalized with 10% does _not_ mean anymore that the validators would have to forfeit 10% of the stake to finalize a conflicting block, if they can just ignore Highway and create finality signatures for something else.

Let's discuss next Wednesday.